### ISSUE TITLE
TempEntities in S2

### ISSUE THREAD

**Issue by fischerq** [skadistats/clarity#72]
I believe the TempEntities got completely removed in S2 replays (or to be precised, moved into DataUserMessages).
It might be good to remove the TempEntity processor, or at least indicate that it is deprecated.

References:
* I ran into the issue when I tried to use the [OnTempEntity](https://github.com/skadistats/clarity/blob/master/src/main/java/skadistats/clarity/processor/tempentities/OnTempEntity.java) event and it did not fire at all.
* The net messages for TempEntities packages got removed from [S1](https://github.com/skadistats/clarity-protobuf/blob/1843156172691809c0f5214fb328866844b4f509/src/main/java/skadistats/clarity/wire/s1/proto/S1NetMessages.java#L250-L251) to [S2](https://github.com/skadistats/clarity-protobuf/blob/1843156172691809c0f5214fb328866844b4f509/src/main/java/skadistats/clarity/wire/s2/proto/S2NetMessages.java#L231)
* New entries in the [DotaUserMessages](https://github.com/skadistats/clarity-protobuf/blob/1843156172691809c0f5214fb328866844b4f509/src/main/java/skadistats/clarity/wire/s2/proto/S2DotaUserMessages.java#L797-L801)

**Comment by spheenik** [comment]
You're right, I figured instead of deprecating, I'd better supply the new ones with a common interface (like combat log). Just have not come around to doing it yet.

Is there any useful information to get from temp entities?

**Comment by fischerq** [comment]
Temp entities are primarily used for projectiles and animations. I'm planning to used it for quite a lot, i.e. timing (and targeting) of attacks, animation canceling, dodging of hits/spells.

I haven't looked into the specific fields available in the S2 ones, but I guess it didn't change much.

**Comment by spheenik** [comment]
Hi @fischerq,

ten years late again 😄 — but the picture in Source 2 turned out to be more interesting than just "moved into DataUserMessages". Here's what we found while looking into this for clarity 3.1.4-SNAPSHOT:

**S2 temp entities actually live in three places:**

1. **Engine-level `te-s2.proto`** (`S2TempEntities` in clarity-protobuf, message IDs 400–426). This file defines 27 typed `CMsgTE*` messages — `CMsgTEEffectDispatch`, `CMsgTEExplosion`, `CMsgTEBeamPoints`, etc. — and is wired up for **all three** Source 2 games (Dota 2, CS2, Deadlock). In practice Valve collapses ~99% of traffic through the polymorphic `CMsgTEEffectDispatch`, with the actual handler selected by an `effectname` uint32 that indexes into the `EffectDispatch` string table (typical entries: `ParticleEffect`, `csblood`, `Impact`, `KnifeSlash`, `ParticleEffectStop`).

2. **Dota-specific `CDOTAUserMsg_TE_*`** in `dota_usermessages.proto`. Six typed messages: `CDOTAUserMsg_TE_Projectile`, `CDOTAUserMsg_TE_ProjectileLoc`, `CDOTAUserMsg_TE_DestroyProjectile`, `CDOTAUserMsg_TE_DotaBloodImpact`, `CDOTAUserMsg_TE_UnitAnimation`, `CDOTAUserMsg_TE_UnitAnimationEnd`. These are direct successors of the DT-class-based S1 Dota temp entities (`DT_TEDOTAProjectile` etc.) — the migration from S1 to S2 was 1:1, just turned into protobufs.

3. **The original S1 path** (`CSVCMsg_TempEntities`, the existing `TempEntities` processor) is still there for S1 Dota 2 replays. This was actually broken in a sneaky way: the `@Provides` engine list was `{CSGO_S1, CSGO_S1}` — a duplicated typo that scoped the processor to CSGO instead of Dota, where it then crashed because CSGO uses a different bit format for temp entities. Fixed in skadistats/clarity@f4b375f.

**Good news for users:** clarity already handles all of this through the existing `@OnMessage` mechanism — no new processor or API was needed. The only piece that's not obvious is the `EffectDispatch` string-table resolution, since the link between the `effectname` integer and the human-readable handler name lives entirely in the replay's stringtables packet.

**Three new examples** in `clarity-examples` demonstrate each path with the same pattern (build a histogram of what was observed):

- [`s1dotatempentities`](https://github.com/skadistats/clarity-examples/tree/master/src/main/java/skadistats/clarity/examples/s1dotatempentities) — Source 1 Dota 2, via the existing `OnTempEntity` event
- [`s2effectdispatch`](https://github.com/skadistats/clarity-examples/tree/master/src/main/java/skadistats/clarity/examples/s2effectdispatch) — Source 2 engine `CMsgTEEffectDispatch`, with the `EffectDispatch` stringtable lookup bundled as a custom `@OnEffectDispatch` event
- [`s2dotatempentities`](https://github.com/skadistats/clarity-examples/tree/master/src/main/java/skadistats/clarity/examples/s2dotatempentities) — Source 2 Dota 2 specific `CDOTAUserMsg_TE_*` messages, consumed directly with `@OnMessage` on the typed protobuf classes

For your original use case (`@OnTempEntity` not firing on S2 replays): that's correct and expected — the S1 processor is now scoped to S1 only, and on S2 you use the typed `@OnMessage` paths shown in the examples above. No deprecation banner on the S1 processor either; it's still the right answer for S1 Dota 2 replays.

**Known gap:** CSGO Source 1 temp entities use a wire format that's distinct from both Dota S1 and the regular CSGO packet-entity-update format, and clarity doesn't currently decode them. That's a separate concern from this issue and should probably get its own ticket if anyone needs it.

Closing this — happy to reopen if anything's missing.
